[EXIR] Register _clone_dim_order op and map aten.clone

keyprocedure opened the pull request:
### Summary
This is PR 2 of 3 implementing a dim order aware clone op. 

This PR registers the new `_clone_dim_order` op and maps `aten.clone` ops to `dim_order_ops._clone_dim_order` in EXIR during export to preserve memory layout changes (contiguous/channels_last). It also updates Core ML and ARM backends to handle the new clone op.

Related PRs:
- PR 1: [#12974](https://github.com/pytorch/executorch/pull/12974) - Add `_clone_dim_order` portable kernel
- PR 3: [#12976](https://github.com/pytorch/executorch/pull/12976) -  Update RemoveCloneOpsTransform to be dim order aware

Fixes #12645  

### Test plan
- Operator level tests to verify kernel behavior for layout preservation and changes.
- Graph level checks to confirm that clone mapping occurs.
- End to end tests to validate that functional clone behavior is unchanged.

All tests pass via:
`python -m unittest exir.tests.test_memory_format_ops_pass`
`python -m unittest backends.apple.coreml.test.test_torch_ops`
`pytest backends/arm/test/ops/test_clone.py`
`pytest backends/arm/test/passes/test_remove_clone_pass.py`

keyprocedure committed:
Register clone_dim_order op; add test for op replacement

keyprocedure committed:
Rename clone_dim_order op registration test

Gasoonjia reviewed (commented):
LGTM! Left a minor feedback about test coverage!

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-aot

keyprocedure commented:
@pytorchbot label "release notes: none"

keyprocedure commented:
@Gasoonjia splitting the `clone_dim_order` op implementation into 3 PRs was a great suggestion, it makes it much easier to isolate and debug issues.

The CI failures are mainly due to two issues:
1.  `clone` is now mapped to `clone_dim_order`, so tests expecting `executorch_exir_dialects_edge__ops_aten_clone_default` are failing because they now see `executorch_exir_dialects_edge__ops_dim_order_ops__clone_dim_order_default`.
2. Because of the first issue, the lack of a kernel implementation for `clone_dim_order` in this PR is also triggering failures. 

Do you have any suggestions for how we should handle the current approach of mapping `clone` to `clone_dim_order`?

Gasoonjia commented:
Hi @keyprocedure thanks for your update!

For places checking the `executorch_exir_dialects_edge__ops_aten_clone_default`, please update them to check `executorch_exir_dialects_edge__ops_dim_order_ops__clone_dim_order_default ` as long as the `compile_config.skip_dim_order == False` in their pipeline.

> the lack of a kernel implementation for clone_dim_order

can you share me more insights here? I think you've create clone_dim_order kernel in aot side by reusing aten clone op.

keyprocedure commented:
@Gasoonjia during export `_clone_dim_order` does fallback to use ATen's `clone` implementation, but at runtime the graph contains `_clone_dim_order` and since no kernel exists for it, it fails:
`[operator_registry.cpp:256] kernel 'dim_order_ops::_clone_dim_order.out' not found.`

This is also why I placed the tests using `memory_format_test_runner` in the kernel PR instead of this one. 

What are your thoughts?

Gasoonjia commented:
@keyprocedure  thanks for detailed reply!

let's swap the order of PRs. We can first finish the runtime op, then aot replacement, finally RemoveCloneOpsTransform.
That would solve the runtime operator requirement issue.

keyprocedure commented:
> @keyprocedure thanks for detailed reply!
> 
> let's swap the order of PRs. We can first finish the runtime op, then aot replacement, finally RemoveCloneOpsTransform. That would solve the runtime operator requirement issue.

That's a great idea! Should I move the tests from the runtime/kernel PR (since the new op won't be registered yet) to the aot replacement PR?

Gasoonjia commented:
> > @keyprocedure thanks for detailed reply!
> > let's swap the order of PRs. We can first finish the runtime op, then aot replacement, finally RemoveCloneOpsTransform. That would solve the runtime operator requirement issue.
> 
> That's a great idea! Should I move the tests from the runtime/kernel PR (since the new op won't be registered yet) to the aot replacement PR?

Yes please! More specific:

for the runtime pr, it should only contain:

1. the portable clone_dim_order op
2. the operator test itself

for aot pt, it should contains both graph-level check, and the end2end test using pybinding.

keyprocedure committed:
Add graph level and end to end tests for _clone_dim_order op

keyprocedure committed:
Remove _clone_dim_order op registration (moved to PR #12974)

Gasoonjia commented:
Thanks for sharing me the error msg!
I've left some comment for the kernel PR, please take a look.

For the issue in this PR, can you share me the code pointer to the issue? Like which line raising that?

keyprocedure commented:
<details>

<summary>There were a total of 49 failing tests for the initial CI run, so I listed out the test and line number of each failure:</summary>

### unittest-arm-backend-with-no-fvp (test_pytest_ops) / linux-job
19 failures
```
backends/arm/test/models/stable_diffusion/test_T5EncoderModel.py:100:
    TestT5EncoderModel.test_T5EncoderModel_tosa_BI

"AssertionError: Invalid partition, found dependency cycles"
```
```
backends/arm/test/models/stable_diffusion/test_T5EncoderModel.py:80:
    TestT5EncoderModel.test_T5EncoderModel_tosa_MI

"IndexError: list index out of range"
```
```
backends/arm/test/models/stable_diffusion/test_CLIPTextModelWithProjection.py:76:
    TestCLIPTextModelWithProjection.test_CLIPTextModelWithProjection_tosa_MI

backends/arm/test/models/stable_diffusion/test_SD3Transformer2DModel.py:128:
    TestSD3Transformer2DModel.test_SD3Transformer2DModel_tosa_BI

backends/arm/test/models/stable_diffusion/test_SD3Transformer2DModel.py:105:
    TestSD3Transformer2DModel.test_SD3Transformer2DModel_tosa_MI

backends/arm/test/models/stable_diffusion/test_vae_AutoencoderKL.py:74:
    TestAutoencoderKL.test_AutoencoderKL_tosa_BI

backends/arm/test/models/stable_diffusion/test_vae_AutoencoderKL.py:55:
    TestAutoencoderKL.test_AutoencoderKL_tosa_MI

backends/arm/test/models/test_conformer.py:60:
    test_conformer_tosa_MI

backends/arm/test/models/test_deit_tiny_arm.py:45:
    test_deit_tiny_tosa_MI

backends/arm/test/models/test_deit_tiny_arm.py:58:
    test_deit_tiny_tosa_BI

backends/arm/test/models/test_dl3_arm.py:44:
    test_dl3_tosa_MI

backends/arm/test/models/test_dl3_arm.py:57:
    test_dl3_tosa_BI

backends/arm/test/models/test_mobilenet_v2_arm.py:45:
    test_mv2_tosa_MI

backends/arm/test/models/test_mobilenet_v2_arm.py:60:
    test_mv2_tosa_BI[per_channel_quantization=true]
    test_mv2_tosa_BI[per_channel_quantization=false]

backends/arm/test/models/test_mobilenet_v2_arm.py:78:
    test_mv2_u55_BI[per_channel_quantization=true]
    test_mv2_u55_BI[per_channel_quantization=false]

backends/arm/test/models/test_mobilenet_v2_arm.py:96:
    test_mv2_u85_BI[per_channel_quantization=true]
    test_mv2_u85_BI[per_channel_quantization=false]

"RuntimeError: Expected to not find "torch.ops.higher_order.executorch_call_delegate" but found it"
```

The CI log also shows multiple TOSA partitioning rejections due to _clone_dim_order not being in BaseTOSASupportList.


***
### unittest-arm-backend-with-no-fvp (test_pytest_models) / linux-job
30 failures
```
backends/arm/test/ops/test_alias_copy.py:63:
    test_alias_tosa_BI[1d_ramp]
    test_alias_tosa_BI[2d_ones]
    test_alias_tosa_BI[3d_rand]
    test_alias_tosa_BI[4d_zeros]

"RuntimeError: Expected out tensor to have dtype signed char, but got float instead"
```

```
backends/arm/test/ops/test_clone.py:58:
    test_clone_tosa_MI[ones_1D_10]
    test_clone_tosa_MI[ones_1D_50]
    test_clone_tosa_MI[rand_1D_20]
    test_clone_tosa_MI[rand_2D_10x10]
    test_clone_tosa_MI[rand_3D_5x5x5]
    test_clone_tosa_MI[rand_4D_2x3x4x5]
    test_clone_tosa_MI[large_tensor]

backends/arm/test/ops/test_clone.py:69:
    test_clone_tosa_BI[ones_1D_10]
    test_clone_tosa_BI[ones_1D_50]
    test_clone_tosa_BI[rand_1D_20]
    test_clone_tosa_BI[rand_2D_10x10]
    test_clone_tosa_BI[rand_3D_5x5x5]
    test_clone_tosa_BI[rand_4D_2x3x4x5]
    test_clone_tosa_BI[large_tensor]

backends/arm/test/passes/test_remove_clone_pass.py:43:
    test_remove_clone_tosa_BI

"RuntimeError: Expected to find "torch.ops.higher_order.executorch_call_delegate" but did not find it"
```
```
backends/arm/test/ops/test_multihead_attention.py:48:
    test_multihead_attention_tosa_MI[rand_2d]
    test_multihead_attention_tosa_MI[randn_2d]
    test_multihead_attention_tosa_MI[randn_3d]

backends/arm/test/ops/test_multihead_attention.py:65:
    test_multihead_attention_tosa_BI[rand_2d]
    test_multihead_attention_tosa_BI[randn_2d]
    test_multihead_attention_tosa_BI[randn_3d]

backends/arm/test/ops/test_multihead_attention.py:108:
    test_multihead_attention_u85_BI[rand_2d]
    test_multihead_attention_u85_BI[randn_2d]
    test_multihead_attention_u85_BI[randn_3d]

backends/arm/test/ops/test_repeat.py:74:
    test_repeat_tosa_MI[interleave_int_3_x_1]

"RuntimeError: Expected to not find "torch.ops.higher_order.executorch_call_delegate" but found it"
```
```
backends/arm/test/ops/test_repeat.py:86:
    test_repeat_tosa_BI[interleave_int_3_x_1]

"RuntimeError: Expected out tensor to have dtype signed char, but got float instead"
```
</details>

keyprocedure commented:
I think we should land the kernel PR before addressing any new failures from the latest CI run in this PR since this PR is dependent on the kernel existing for `_clone_dim_order`.

Gasoonjia commented:
sure let's land prs one by one

keyprocedure committed:
Register _clone_dim_order op

keyprocedure committed:
Merge branch 'main' into add-dim-order-clone-aot

Gasoonjia reviewed (commented):
LGTM thanks for your wonderful work!
Will stamp the PR after ci passed!

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-aot

keyprocedure commented on exir/tests/test_memory_format_ops_pass.py:
```diff
@@ -153,6 +205,29 @@ def test_op_dim_order_propagation_ambiguous(self) -> None:
         except AmbiguousDimOrderError:
             pass  # Expected error
 
+    def test_op_clone_dim_order_graph_replacement(self):
```
Oh okay, that makes sense. I'll remove this test since the other tests I added use `memory_format_test_runner`, which calls `to_edge` and `_check_ir_validity` defaults to True.

keyprocedure committed:
Register _clone_dim_order as no-op in CoreML

keyprocedure committed:
Remove redundant _clone_dim_order graph check

keyprocedure committed:
Add _clone_dim_order to RemoveClonePass and update op name in tests

keyprocedure committed:
Register _clone_dim_order under TOSA support check

keyprocedure commented:
@Gasoonjia 
Could you take a look at the changes I made to fix the ARM failures from the last CI run when you have a chance? I added `_clone_dim_order` to the TOSA support check and updated the tests. Is TOSA available publicly? I couldn’t run the tests locally because I couldn’t source the package.

digantdesai commented on backends/apple/coreml/compiler/torch_ops.py:
```diff
@@ -67,6 +68,28 @@ def _to_dim_order_copy(context, node):
+    ), "Only contiguous memory format is supported in CoreML"
+
+    # Since CoreML only supports contiguous format, no dim_order preservation is needed. Treat this as a no-op clone.
+    noop(context, node)
```
Don't we have to `clone`?

keyprocedure replied:
`clone` is in the `torch list` for the noop handler, so I assumed that during conversion all cloning was simply returning the input. Does that seem right, or am I missing something?

digantdesai commented on backends/arm/_passes/remove_clone_pass.py:
```diff
@@ -14,7 +14,11 @@ class RemoveClonePass(ExportPass):
     def call_operator(self, op, args, kwargs, meta):
-        if op != exir_ops.edge.aten.clone.default:
+        clone_ops = (
+            exir_ops.edge.aten.clone.default,
```
why do we need `aten.clone` we should be removing this when producing the edge dialect graph, no?

keyprocedure replied:
Because isn't `aten.clone` only mapped to `_clone_dim_order` if `_skip_dim_order=False`?

keyprocedure replied:
But that's a good point. Can we assume `_skip_dim_order` is always `False` in ARM? If so, we can replace all instances of `aten.clone` with `_clone_dim_order`. If not, we'd need to make the clone tests check for the correct clone op based on the flag's value. I also assumed that all clones are removed in ARM regardless of whether the dim order changes, so I didn’t add a dim_order check in the removal pass.

digantdesai commented on backends/arm/operator_support/clone_dim_order_support.py:
```diff
@@ -0,0 +1,76 @@
+
+
+@register_tosa_support_check
+class CloneDimOrderSupport(SupportedTOSAOperatorCheck):
```
Did we remove support to partition aten.clone op?

keyprocedure replied:
We didn't, `aten.clone` is still in the `BaseTOSASupportList` and supported. `CloneDimOrderSupport` would only apply when cloning takes place with `_skip_dim_order=False`, otherwise `aten.clone` would still exist.

digantdesai commented on backends/apple/coreml/compiler/torch_ops.py:
```diff
@@ -67,6 +68,28 @@ def _to_dim_order_copy(context, node):
     to(context, node)
 
 
+@register_torch_op(
```
did we remove support to partition `aten.clone` op?

keyprocedure replied:
I didn't explicitly remove support for `aten.clone`. From my understanding this change would only add support for `_clone_dim_order`, `aten.clone` is still supported and handled by the noop handler in `coremltools`.

digantdesai reviewed (commented):
Looks good for the most part. Left some comments.

digantdesai committed:
Merge branch 'main' into add-dim-order-clone-aot

keyprocedure committed:
Add clone_dim_order_support to TOSA operator support list

keyprocedure committed:
Register node visitor for _clone_dim_order

Gasoonjia reviewed (commented):
LGTM, though there're some Arm question maybe we need to wait for @digantdesai  for final decision.

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-aot

keyprocedure commented on backends/arm/test/misc/test_partition_decomposed_quantized_ops.py:
```diff
@@ -38,7 +38,7 @@
 linear_residual_exir_op: list[str] = [
     "executorch_exir_dialects_edge__ops_aten_gelu_default",
-    "executorch_exir_dialects_edge__ops_aten_clone_default",
+    "executorch_exir_dialects_edge__ops_dim_order_ops__clone_dim_order_default",
```
Since ARM's pass manager doesn't seem to pass the `_skip_dim_order` flag, it defaults to `False`, so my strategy was to replace all instances of `aten.clone` with `_clone_dim_order` in the tests. Although it might make more sense to explicitly set `_skip_dim_order=True` in the `EdgeCompileConfig` calls in [arm_tester](https://github.com/pytorch/executorch/blob/1164fe4100cd82ab3c7dbfa2a0589fe6386d27d0/backends/arm/test/tester/arm_tester.py#L160) and undo my changes to the ARM test files. Then I can either continue with adding `_clone_dim_order` support in TOSA for future use or leave it out for now.

digantdesai replied:
FYI see - https://github.com/pytorch/executorch/commit/135e875f2f0c9ee9704899c1208512d5fcf697bd

digantdesai commented on backends/apple/coreml/compiler/torch_ops.py:
```diff
@@ -67,6 +68,28 @@ def _to_dim_order_copy(context, node):
     to(context, node)
 
 
+@register_torch_op(
```
FYI @metascroy

metascroy replied:
This change looks OK to me, but do we have a test that covers it?

oscarandersson8218 reviewed (commented):
@keyprocedure Nice change and thanks for updating our backend!

keyprocedure commented on backends/arm/operators/op_clone_dim_order.py:
```diff
@@ -0,0 +1,43 @@
+
+
+@register_node_visitor
+class CloneDimOrderVisitor(NodeVisitor):
```
Thanks for pointing this out, I'll remove this NodeVisitor. Just to confirm, if a clone changes dim_order, should the RemoveClonePass still be expected to remove it?

Gasoonjia commented:
hi @keyprocedure looks like there're still several issues on ci. Do you mind take a look?

keyprocedure commented:
> hi @keyprocedure looks like there're still several issues on ci. Do you mind take a look?

Hi, yeah, currently working on them. Appreciate the check in :)

keyprocedure committed:
Remove visitor node registration for _clone_dim_order

keyprocedure committed:
Remove aten.clone check from RemoveClonePass

keyprocedure committed:
Remove input dtype gating and add memory_format check

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-aot

keyprocedure commented:
@Gasoonjia it looks like we're good! 

The 3 CI failures seem unrelated:
- `build doc (buck2)` failed because of an upload error
- `test-static-llama-qnn-linux` had a `Failed to download LLVM` error
- `test-llama_runner_eager-linux` failed without a log (GitHub suggests `FLAKY - The following job failed but was likely due to flakiness present on trunk`)

metascroy commented on backends/apple/coreml/compiler/torch_ops.py:
```diff
@@ -67,6 +68,28 @@ def _to_dim_order_copy(context, node):
+    ), "Only contiguous memory format is supported in CoreML"
+
+    # Since CoreML only supports contiguous format, no dim_order preservation is needed. Treat this as a no-op clone.
+    noop(context, node)
```
Why not just call clone(context, node) here?

In case coremltools updates clone to be something other than a noop in future?

metascroy replied:
I guess this is OK then

keyprocedure committed:
Add Core ML test for _clone_dim_order

Gasoonjia commented:
LGTM. Thanks for your contribution! @keyprocedure 

 @digantdesai  @oscarandersson8218 @metascroy please let me know if you have any concerns, or I will stamp it.

keyprocedure commented:
Appreciate all the feedback and reviews, glad we got this landed!

jackzhxng commented:
Sorry to be the bearer of bad news, but this broke some trunk tests. Let's revert and @keyprocedure can you try to get this in again with the `ciflow/trunk` label turned on?

Gasoonjia commented:
@jackzhxng  stamped the revert PR. 
btw why do we need to manually set the tag for ci test, instead of make it as default? 
is there any place we can find all tags that should be tagged for enabling all cis?

keyprocedure commented:
Thanks for the update @jackzhxng, I'll resubmit the PR with the additional label.

digantdesai commented:
> btw why do we need to manually set the tag for ci test, instead of make it as default?
money 🤑
